### ISSUE TITLE
SNOW-822793: A deadlock on connection object release

### ISSUE THREAD

**Issue by asiunov** [snowflakedb/snowflake-connector-python#1567]
1. What version of Python are you using?
```
Python 3.9.12 (main, Mar 25 2022, 00:41:03) 
[GCC 8.3.0]
```

2. What operating system and processor architecture are you using?
```
Linux-5.4.0-1105-azure-x86_64-with-glibc2.28
```

3. What are the component versions in the environment (`pip freeze`)?
```
grpcio==1.53.0
grpcio-status==1.47.0
grpcio-tools==1.39.0
snowflake-connector-python==2.8.3
snowflake-sqlalchemy==1.4.5
...
```

4. What did you do?

I run a service with many components inside. One of which runs `ThreadPoolExecutor`. I noticed that it is stuck processing anything. Here is the threaddump for this thread:
```
  File "/myapp.py", line 13, in run
    executor.submit(func)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 162, in submit
    with self._shutdown_lock, _global_shutdown_lock:
```
I found another thread with this threaddump:
```
  File "/usr/local/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()
  File "/usr/local/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.9/site-packages/grpc/_server.py", line 1004, in _serve
    if not _process_event_and_continue(state, event):
  File "/usr/local/lib/python3.9/site-packages/grpc/_server.py", line 968, in _process_event_and_continue
    rpc_state, rpc_future = _handle_call(event, state.generic_handlers,
  File "/usr/local/lib/python3.9/site-packages/grpc/_server.py", line 856, in _handle_call
    return _handle_with_method_handler(rpc_event, method_handler,
  File "/usr/local/lib/python3.9/site-packages/grpc/_server.py", line 829, in _handle_with_method_handler
    return state, _handle_unary_unary(rpc_event, state,
  File "/usr/local/lib/python3.9/site-packages/grpc/_server.py", line 716, in _handle_unary_unary
    return thread_pool.submit(_unary_response_in_pool, rpc_event, state,
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 176, in submit
    self._adjust_thread_count()
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 182, in _adjust_thread_count
    if self._idle_semaphore.acquire(timeout=0):
  File "/usr/local/lib/python3.9/threading.py", line 432, in acquire
    with self._cond:
  File "/usr/local/lib/python3.9/site-packages/snowflake/connector/connection.py", line 305, in __del__
    self.close(retry=False)
  File "/usr/local/lib/python3.9/site-packages/snowflake/connector/connection.py", line 567, in close
    if self._all_async_queries_finished():
  File "/usr/local/lib/python3.9/site-packages/snowflake/connector/connection.py", line 1576, in _all_async_queries_finished
    for f in as_completed(futures):
  File "/usr/local/lib/python3.9/concurrent/futures/_base.py", line 222, in as_completed
    fs = set(fs)
  File "/usr/local/lib/python3.9/site-packages/snowflake/connector/connection.py", line 1575, in <genexpr>
    futures = (tpe.submit(async_query_check_helper, sfqid) for sfqid in queries)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 162, in submit
    with self._shutdown_lock, _global_shutdown_lock:
```
If I understand correctly what's going on here is:
this is one of the threads of my gRPC server (I run this in addition to the thread in the first stacktrace), which uses `ThreadPoolExecutor`. It locks `_global_shutdown_lock` as part of `thread_pool.submit` (because `self._adjust_thread_count()` runs inside `with self._shutdown_lock, _global_shutdown_lock:` block). This lock has type `threading.Lock` (not `RLock`, so it does not allow double lock in the same thread). Then, at some point, python decided to release a snowflake connection object in this thread, i.e. the gRPC code pauses while the GC happens in this thread. Then, as part of `snowflake.Connection.__del__`, it calls `.close`, which uses `ThreadPoolExecutor` to verify whether all queries are completed (`self._all_async_queries_finished`), which is trying to lock on `with self._shutdown_lock, _global_shutdown_lock:`, i.e. it is deadlocked due to the double lock on `_global_shutdown_lock` in the same thread.

5. What did you expect to see?

No deadlock on connection close.

6. Can you set logging to DEBUG and collect the logs?

This behavior is hard to replicate in my environment, but I believe it is possible using code like:
```
with _global_shutdown_lock: # not a prod code, but just to replicate the logic...
  snowflake_connection.close()
```

Thank you for your help!

**Comment by sfc-gh-aling** [comment]
thanks, this is an interesting case, we will take a look into this.
in the meanwhile would it be helpful to mitigate if you proactively close the connection before GC kicks in?

**Comment by asiunov** [comment]
Thanks for looking into this @sfc-gh-aling !

I thought about this mitigation but I'm actually not sure if this would help, because, if I read the snowflake connection code correctly, `close` method considers all queries in `self._async_sfqids` dict but it does not remove queries that are still running, so even if I close manually, it would not help because `__del__` would check these queries again.

**Comment by asiunov** [comment]
I just got another very similar issue, that is deadlocked within snowflake on a different global lock object:
```
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/result_batch.py", line 698, in create_iter
  return self._get_arrow_iter(connection=connection)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/result_batch.py", line 646, in _get_arrow_iter
  return self._create_iter(iter_unit=IterUnit.TABLE_UNIT, connection=connection)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/result_batch.py", line 634, in _create_iter
  response = self._download(connection=connection)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/result_batch.py", line 297, in _download
  response = session.request("get", **request_data)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/requests/sessions.py", line 587, in request
  resp = self.send(prep, **send_kwargs)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/requests/sessions.py", line 701, in send
  r = adapter.send(request, **kwargs)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/requests/adapters.py", line 489, in send
  resp = conn.urlopen(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/urllib3/connectionpool.py", line 703, in urlopen
  httplib_response = self._make_request(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/urllib3/connectionpool.py", line 386, in _make_request
  self._validate_conn(conn)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/urllib3/connectionpool.py", line 1042, in _validate_conn
  conn.connect()
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/urllib3/connection.py", line 414, in connect
  self.sock = ssl_wrap_socket(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ssl_wrap_socket.py", line 86, in ssl_wrap_socket_with_ocsp
  v = SFOCSP(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 985, in validate
  return self._validate(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 998, in _validate
  results = self._validate_certificates_sequential(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 1167, in _validate_certificates_sequential
  self._check_ocsp_response_cache_server(cert_data)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 1236, in _check_ocsp_response_cache_server
  self.OCSP_CACHE_SERVER.download_cache_from_server(self)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 353, in download_cache_from_server
  retval = OCSPServer._download_ocsp_response_cache(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 408, in _download_ocsp_response_cache
  ocsp.decode_ocsp_response_cache(response.json())
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 1597, in decode_ocsp_response_cache
  found, _ = OCSPCache.find_cache(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 718, in find_cache
  ) and ocsp.is_valid_time(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_asn1crypto.py", line 225, in is_valid_time
  if basic_ocsp_response["certs"].native:
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 4525, in native
  self._parse_children(recurse=True)
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 4491, in _parse_children
  child._parse_children(recurse=True)
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 3962, in _parse_children
  child._parse_children(recurse=True)
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 3962, in _parse_children
  child._parse_children(recurse=True)
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 4491, in _parse_children
  child._parse_children(recurse=True)
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 3895, in _parse_children
  cls._precomputed_specs[field] or self._determine_spec(field))
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 3751, in _determine_spec
  oid = self._lazy_child(self._oid_nums[0]).native
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 3194, in native
  self._native = self.dotted
File "/usr/local/lib/python3.9/site-packages/asn1crypto/core.py", line 3152, in dotted
  output = []
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/connection.py", line 305, in __del__
  self.close(retry=False)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/connection.py", line 569, in close
  self.rest.delete_session(retry=retry)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/network.py", line 621, in delete_session
  ret = self._post_request(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/network.py", line 704, in _post_request
  ret = self.fetch(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/network.py", line 794, in fetch
  ret = self._request_exec_wrapper(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/network.py", line 837, in _request_exec_wrapper
  return_object = self._request_exec(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/network.py", line 1018, in _request_exec
  raw_ret = session.request(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/requests/sessions.py", line 587, in request
  resp = self.send(prep, **send_kwargs)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/requests/sessions.py", line 701, in send
  r = adapter.send(request, **kwargs)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/requests/adapters.py", line 489, in send
  resp = conn.urlopen(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/urllib3/connectionpool.py", line 703, in urlopen
  httplib_response = self._make_request(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/urllib3/connectionpool.py", line 386, in _make_request
  self._validate_conn(conn)
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/urllib3/connectionpool.py", line 1042, in _validate_conn
  conn.connect()
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/vendored/urllib3/connection.py", line 414, in connect
  self.sock = ssl_wrap_socket(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ssl_wrap_socket.py", line 86, in ssl_wrap_socket_with_ocsp
  v = SFOCSP(
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 919, in __init__
  if not OCSP_RESPONSE_VALIDATION_CACHE:
File "/usr/local/lib/python3.9/site-packages/snowflake/connector/cache.py", line 65, in __len__
  with self._lock:
```

This frame `File "/usr/local/lib/python3.9/site-packages/snowflake/connector/ocsp_snowflake.py", line 1597, in decode_ocsp_response_cache` is locked on `with OCSP_RESPONSE_VALIDATION_CACHE._lock:` which is a global object in snowflake library, it is followed by an interruption caused by `snowflake.Connection.__del__` and the last frame is locked on the same `OCSP_RESPONSE_VALIDATION_CACHE._lock` object, which is `threading.Lock()`.

**Comment by sfc-gh-dszmolka** [comment]
we hope this issue to be addressed in https://github.com/snowflakedb/snowflake-connector-python/pull/1898

**Comment by sfc-gh-dszmolka** [comment]
fix released with v3.8.0, thank you for bearing with us !
